allennlp/models/language_model.py

```diff
@@ -0,0 +1,206 @@
+                 encoder: Seq2SeqEncoder,
+                 softmax: SoftmaxWithNLL,
+                 proj: bool = False,
+                 relu: bool = False,
+                 dropout: float = 0.0,
+                 batch_first: bool=True,
+                 initializer: InitializerApplicator = InitializerApplicator(),
+                 regularizer: Optional[RegularizerApplicator] = None) -> None:
+
+        super().__init__(vocab, regularizer)
+
+        self.encoder = encoder
+        self.batch_first = batch_first
+        self.text_field_embedder = text_field_embedder
+
+        if dropout:
+            self.dropout = torch.nn.Dropout(dropout)
+            dropout_list = [self.dropout]
+        else:
+            self.dropout = None
+            dropout_list = []
+
+        if proj:
+            self.proj = nn.Linear(self.encoder.get_output_dim(), softmax.get_input_dim())
+            proj_list = [self.proj] 
+            if relu:
+                proj_list += [nn.ReLU()]
+            proj_list = proj_list + dropout_list
+        else:
+            self.proj = None
+            proj_list = []
+
+        after_encoder_list = dropout_list + proj_list
+        if len(after_encoder_list) > 0:
+            self.after_encoder = nn.Sequential(*after_encoder_list)
+        else:
+            self.after_encoder = None
+
+        self.softmax = softmax
+        if not self.softmax.adaptive and self.softmax.head.weight.size() == self.text_field_embedder.token_embedder_tokens.weight.size():
+            self.softmax.head.weight = self.text_field_embedder.token_embedder_tokens.weight
+
+        self.softmax_in = softmax.get_input_dim()
+
+        self.metrics = {"ppl": Perplexity()}
+
+        check_dimensions_match(text_field_embedder.get_output_dim(), 
+            encoder.get_input_dim(), 
+            "text field embedder output dim",
+            "encoder input dim")
+        if not proj:
+            check_dimensions_match(encoder.get_output_dim(), 
+                self.softmax_in,
+                "encoder output dim",
+                "softmax input dim")
+
+        initializer(self)
+
+    @overrides
+    def get_metrics(self, reset: bool = False) -> Dict[str, float]:
+        return {metric_name: metric.get_metric(reset) for metric_name, metric in self.metrics.items()}
+        
+    @overrides
+    def train(self, mode=True):
+        self.encoder.reset_states()
+        self.training = mode
+        for module in self.children():
+            module.train(mode)
+        return self
+
+    @overrides
+    def eval(self):
+        return self.train(False)
+
+    @overrides
+    def forward(self,  # type: ignore
+                input_tokens: Dict[str, torch.LongTensor],
+                output_tokens: Dict[str, torch.LongTensor] = None) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ
+        """
+        Parameters
+        ----------
+        input_tokens : Dict[str, torch.LongTensor], required
```

Comment:
Docstring should contain something module specific and should remove the standard template.

Reply:
@LiyuanLucasLiu Just wondering if you are planning to complete this PR. Would be very useful.